### PR TITLE
Jit64: Merge memory allocations into a single allocation

### PR DESCRIPTION
**Pull request by MerryMage**
As promised in #5110.

Instead of each component allocating their own memory, we instead allocate
the memory once and divvy that up among the components as required. This
ensures that relative memory offsets remain within architecture limits.

This was done by extending the pre-existing concept of a child codeblock
to support multiple children.

### CODE REVIEW

File: Source/Core/Core/PowerPC/Jit64/Jit.cpp [hunk 1]
```diff
@@ -229,8 +229,15 @@ void Jit64::Init()
...
   fpr.SetEmitter(this);
 
-  trampolines.Init(jo.memcheck ? TRAMPOLINE_CODE_SIZE_MMU : TRAMPOLINE_CODE_SIZE);
-  AllocCodeSpace(CODE_SIZE);
+  const size_t routines_size = asm_routines.CODE_SIZE;
+  const size_t trampolines_size = jo.memcheck ? TRAMPOLINE_CODE_SIZE_MMU : TRAMPOLINE_CODE_SIZE;
+  const size_t farcode_size = jo.memcheck ? FARCODE_SIZE_MMU : FARCODE_SIZE;
+  const size_t constpool_size = m_const_pool.CONST_POOL_SIZE;
+  AllocCodeSpace(CODE_SIZE + routines_size + trampolines_size + farcode_size + constpool_size);
+  AddChildCodeSpace(&asm_routines, asm_routines.CODE_SIZE);
```
Comment (degasus) [inline Source/Core/Core/PowerPC/Jit64/Jit.cpp]: routines_size

File: Source/Core/Core/PowerPC/Jit64/JitAsm.cpp [hunk 2]
```diff
@@ -20,6 +20,15 @@ using namespace Gen;
 // Not PowerPC state.  Can't put in 'this' because it's out of range...
 static void* s_saved_rsp;
 
+void Jit64AsmRoutineManager::Init(u8* stack_top)
+{
+  m_const_pool.Init(AllocChildCodeSpace(4096), 4096);
+  m_stack_top = stack_top;
+  Generate();
+  WriteProtect();
+  _assert_(GetSpaceLeft() > 0);
```
Comment (degasus) [inline Source/Core/Core/PowerPC/Jit64/JitAsm.cpp]: GetSpaceLeft() is unsigned

File: Source/Core/Core/PowerPC/Jit64IL/JitIL.cpp [hunk 3]
```diff
@@ -262,12 +262,19 @@ void JitIL::Init()
...
   UpdateMemoryOptions();
 
-  trampolines.Init(jo.memcheck ? TRAMPOLINE_CODE_SIZE_MMU : TRAMPOLINE_CODE_SIZE);
-  AllocCodeSpace(CODE_SIZE);
+  const size_t routines_size = asm_routines.CODE_SIZE;
+  const size_t trampolines_size = jo.memcheck ? TRAMPOLINE_CODE_SIZE_MMU : TRAMPOLINE_CODE_SIZE;
+  const size_t farcode_size = jo.memcheck ? FARCODE_SIZE_MMU : FARCODE_SIZE;
+  const size_t constpool_size = m_const_pool.CONST_POOL_SIZE;
+  AllocCodeSpace(CODE_SIZE + routines_size + trampolines_size + farcode_size + constpool_size);
+  AddChildCodeSpace(&asm_routines, asm_routines.CODE_SIZE);
```
Comment (degasus) [inline Source/Core/Core/PowerPC/Jit64IL/JitIL.cpp]: routines_size

File: Source/Core/Common/CodeBlock.h [hunk 4]
```diff
@@ -60,42 +63,55 @@ class CodeBlock : public T, NonCopyable
...
+      child->region = nullptr;
+      child->region_size = 0;
+      child->total_region_size = 0;
     }
   }
 
-  bool IsInSpace(const u8* ptr) const { return ptr >= region && ptr < (region + region_size); }
+  bool IsInSpace(const u8* ptr) const
+  {
+    return ptr >= region && ptr < (region + total_region_size);
```
Comment (degasus) [inline Source/Core/Common/CodeBlock.h]: `region + total_region_size` is out of range for the childs. The old checked it for each region, not for the total region, why did you change it?

File: Source/Core/Common/CodeBlock.h [hunk 5]
```diff
@@ -60,42 +63,55 @@ class CodeBlock : public T, NonCopyable
...
+  }
   // Cannot currently be undone. Will write protect the entire code region.
   // Start over if you need to change the code (call FreeCodeSpace(), AllocCodeSpace()).
-  void WriteProtect() { Common::WriteProtectMemory(region, region_size, true); }
+  void WriteProtect() { Common::WriteProtectMemory(region, total_region_size, true); }
```
Comment (degasus) [inline Source/Core/Common/CodeBlock.h]: This should be also for each child

File: Source/Core/Common/CodeBlock.h (hunk 4 above)
Comment (MerryMage) [inline Source/Core/Common/CodeBlock.h]: I changed the meaning of `region_size`. It originally meant what `total_region_size` means now.

I did that because `parent_region_size` was a confusing name.
